README.md
# SLLM - Command Line ChatGPT-like Assistant
A command line interface for the GPT3 LLM that emulates some features of ChatGPT (Bring your own token).

In addition to providing a simple interface for talking with GPT3, this tool also offers a few extra features built on top of the GPT3 API. 

### Extra Features:
- Act as a chat bot (emulates chatGPT)
- Read local files
- Automatically prepend subject domains & context (Bash, JS, Physics etc.)
- Expert Mode (act as an expert on some subject)
- Explain it like I'm 5 (explain the answer simply)

### Why Not Use ChatGPT?
You can do whatever you want :)

I made this for the following reasons:

- Access GPT3 without leaving the command line
- Access GPT3 without logging in to OpenAI (use a token instead)
- Directly read and write local files
- I don't always have easy access to a GUI

> $ sllm what would be the avantage of talking to a LLM via the command line?
> 
> The advantage of talking to a LLM via the command line is that it allows for a more efficient and direct way of communicating. It also allows for more precise and specific commands to be used, which can help to quickly get the desired results.

---

### Example Usage:

> $ sllm what can be used to check the status of a running systemd service? -e bash scripting
> 
> The command to check the status of a running systemd service is "systemctl status <service_name>".
> 
> 
> $ sllm how can I get the full log instead? -d bash -H 1
> 
> To get the full log of a running systemd service, you can use the command "journalctl -u <service_name>".


## Install

```
npm install -g sllm
```

## Setup

Get an OpenAI [API KEY](https://platform.openai.com/account/api-keys).

```
export OPENAI_API_KEY=<your_api_key>
```

## Quick Start

```
$ sllm how many people live in china? 

According to the latest estimates, there are approximately 1.4 billion people living in China.
```

## Chat Mode
To enable a "chat mode" similar to chatGPT, run the following command:

```
sllm set -H 3
```

This will remind the LLM about the last 3 prompts it was given. 

## Overview
```
$ sllm -h

Usage: sllm [options] [command]

CLI for GPT3. Created by Mathieu Dombrock 2023. GPL3 License.

Options:
  -V, --version                 output the version number
  -h, --help                    display help for command

Commands:
  prompt [options] <prompt...>  send a prompt (default command)
  set [options]                 set a persistant command option
  settings [options]            view the current settings that were changed via the `set` command
  hist [options]                manage the prompt / response history
  purge                         delete all history and settings
  count [options]               estimate the tokens used by a prompt or file
  repeat                        repeat the last response
  help [command]                display help for command

```

## Prompt

```
$ sllm -h prompt

Usage: sllm prompt [options] <prompt...>

Send a prompt (default command)

Arguments:
  prompt                      the prompt text

Options:
  -m, --max-tokens <number>   maximum tokens to use (default: "256")
  -t, --temperature <number>  temperature to use (default: "0.2")
  -c, --context <string...>   context to prepend
  -d, --domain <string...>    subject domain to prepend
  -e, --expert <string...>    act as an expert on this domain
  -5, --like-im-five          explain it like I'm 5 years old
  -H, --history <number>      prepend history (chatGPT mode)
  -f, --file <path>           preprend the given file contents
  -v, --verbose               verbose output
  -M, --mock                  dont actually send the prompt to the API
  -h, --help                  display help for command

```

## Working With Files

You can prepend a reference to a file with the `-f` or `--file` option.

However, be aware that files can not exceed 4k tokens.

```
$ sllm what dependancies does this have -f ./package.json

 This package.json file has two dependencies: gpt-3-encoder and openai.

$ sllm what version of npm is this file built for? -f ./package.json

  This package.json file is built for npm version 6.14.4 or higher.

$ cat example.js

  const e = require('./llm.js');
  console.log(e);

$ sllm is this file NodeJS or Browser JS? -f example.js

  This file is Node.js.

$ sllm why do you say that?

  This file contains code that is specific to Node.js, such as the require statement, which is not supported in browser JavaScript.

$ sllm what is this file about? -f mute.cpp
  This file is about demonstrating the differences between mutating a value by reference, by pointer, and not mutating it at all. It contains three functions, noMute, muteR, and muteP, which respectively do not mutate the value, mutate the value by reference, and mutate the value by pointer. There is also a print function to output the results of the functions.

$ sllm what is this file about? -f cfg.txt
  This file is about creating a GIF animation of Conway's Game of Life using the .sorg settings. The animation will have a file name of "life", a frame delay of 1, 512 frames to render, 0 generations to run before render, a canvas width of 64, a canvas height of 64, a pixel/image scale of 8, a gif color pallet of lime, and a rule set of dtsd. Additionally, the .sorg settings include a file to load of "noise", a center of 0, an x offset of 1, and a y offset of 1.
```

## Counting Tokens

If you want to estimate how many tokens a prompt or file will consume, you can use the `sllm count` command.

```
$ sllm help count

Usage: sllm count [options]

estimate the tokens used by a prompt or file

Options:
  -p, --prompt <string...>  the prompt string to check
  -f, --file <path>         the file path to check
  -h, --help                display help for command
```




JS/history.txt
_user_: How long would that take?

_llm_: 

Establishing a permanent base on the Moon would likely take several years, depending on the resources available and the complexity of the project. It would involve a lot of research and development, and would require the collaboration of many different organizations and countries.



sllm.js
#! /usr/bin/env node

const fs = require('fs');
const os = require('os');
const { encode } = require('gpt-3-encoder');
const { Configuration, OpenAIApi } = require('openai');
const { Command } = require('commander'); // (normal include)

const USER_CFG_DIR = os.homedir() + '/.config/sllm';

const MAX_HISTORY_STORE = 64;

// Read package.json version number
const VERSION_NUMBER = JSON.parse(
	fs.readFileSync(__dirname + '/package.json')
).version;

// Ensure we have an api key env var
_ensureAPIKey();
// Ensure we have our needed files
_ensureFiles();

const configuration = new Configuration({
	apiKey: process.env.OPENAI_API_KEY,
});

const openai = new OpenAIApi(configuration);

async function llm(prompt, options) {
	// Join prompt into single string
	prompt = prompt.join(' ');

	// Set static options
	options = _loadOpts(options);

	// Ensure numbers
	options.maxTokens = Number(options.maxTokens) || 256;
	options.temperature = Number(options.temperature) || 0.2;
	options.history = Number(options.history) || 0;

	// Prepend a file if required
	if (options.file) {
		if (fs.existsSync(options.file)) {
			prompt =
				fs.readFileSync(options.file, 'UTF-8') +
				'\r\n' +
				'Consider the previous file contents when responding to the prompt: ' +
				prompt;
		}
	}

	// Preproceess the prompt
	if (options.likeImFive) {
		prompt = 'Answer this as if I\'m five years old: ' + prompt;
	}
	if (options.context) {
		prompt =
			'In the context of ' + options.context.join(' ') + ', ' + prompt;
	}
	if (options.domain) {
		prompt = 'In the domain of ' + options.domain.join(' ') + ', ' + prompt;
	}
	if (options.expert) {
		prompt =
			'Act as an expert on ' +
			options.expert.join(' ') +
			'. The question I need you to answer is: ' +
			prompt;
	}

	// Append History
	const ogPrompt = prompt; // Cache for history etc
	if (options.history) {
		prompt =
			_loadHistory(options.history, false) + // Revrsed
			prompt;
	}

	// Count total prompt tokens and append to the maxTokens value
	const encoded = encode(prompt);
	const tokenCount = encoded.length;
	const totalTokens = options.maxTokens + tokenCount;

	// Apply verbos outout
	if (options.verbose) {
		console.log('>> ' + ogPrompt + ' <<');
		console.log(JSON.stringify(options, null, 2));
		console.log('Encoded Tokens: ' + encoded.length);
		console.log('Total Potential Tokens: ' + totalTokens);
		console.log('Sending Prompt...');
		console.log('-------\r\n');
	}

	// GPT can use a total of 4096 tokens at once
	// Using 4k for the prompt allows us 96 for response
	if (totalTokens > 4000) {
		console.log('ERROR: Max Tokens Exceeded ' + '(' + totalTokens + ')');
		console.log('Please limit your prompt');
		if (options.history) {
			console.log('Trying running with history off!');
		}
		return;
	}

	// Make the request
	let output = 'WARNING: Did not send!';
	if (!options.mock) {
		const completion = await openai.createCompletion({
			model: 'text-davinci-003',
			prompt: prompt,
			max_tokens: options.maxTokens,
			temperature: options.temperature,
		});
		output = completion.data.choices[0].text;
	}

	// Strip dialog references
	if (options.history) {
		output = output.replace('_user_:', '');
		output = output.replace('_llm_:', '');
	}

	// Check for empty response
	if (output.length < 1) {
		output = 'WARNING: Something went wrong! Try Again.';
	}

	// Trim whitesapce
	output = output.trim();
	// Log output
	console.log(output);

	// Log history
	_logHistory(ogPrompt, output);
}

function history(options) {
	if (options.delete) {
		fs.rmSync(USER_CFG_DIR + '/history.json');
		console.log('Deleted History');
		return;
	}
	// Default to view
	if (options.view) {
		// Ensure number
		options.view = Number(options.view) || 32;
		const content = _loadHistory(options.view, false);
		console.log(content);
		return;
	}
}

function repeat() {
	const last = _loadHistory(1, true, true);
	console.log(last[0].llm);
}

function settings(options) {
	let content;
	if (options.delete) {
		fs.rmSync(USER_CFG_DIR + '/settings.json');
		content = 'Deleted Settings';
	} else {
		content = fs.readFileSync(USER_CFG_DIR + '/settings.json', 'UTF-8');
	}
	console.log(content);
	console.log('Settings can be changed with the `set` command.');
}

function setOpts(options) {
	console.log(JSON.stringify(options));
	fs.writeFileSync(
		USER_CFG_DIR + '/settings.json',
		JSON.stringify(options, null, 2)
	);
	console.log('Created a new settings file');
}

function purge() {
	fs.rmSync(USER_CFG_DIR + '/settings.json');
	fs.rmSync(USER_CFG_DIR + '/history.json');
	console.log('Purged!');
}

function countTokens(options) {
	let tokens = 0;
	if (options.prompt) {
		options.prompt = options.prompt.join(' ');
		const encoded = encode(options.prompt);
		tokens += encoded.length;
	}
	if (options.file) {
		const contents = fs.readFileSync(options.file, 'UTF-8');
		const encoded = encode(contents);
		tokens += encoded.length;
	}
	if (tokens === 0) {
		console.log('ERROR: Nothing to count');
		console.log('Use the --prompt or --file options!');
		return;
	}

	console.log('Estimated Tokens: ' + tokens + '/' + '4096');
	console.log('Max Reply: ' + (4096 - tokens));
}

function _ensureFiles() {
	if (!fs.existsSync(USER_CFG_DIR)) {
		// Create the dir
		fs.mkdirSync(USER_CFG_DIR);
	}
	if (!fs.existsSync(USER_CFG_DIR + '/history.json')) {
		// Create the file
		fs.writeFileSync(USER_CFG_DIR + '/history.json', '[]');
	}
	if (!fs.existsSync(USER_CFG_DIR + '/settings.json')) {
		// Create the file
		fs.writeFileSync(USER_CFG_DIR + '/settings.json', '{}');
	}
}

function _ensureAPIKey() {
	if (!process.env.OPENAI_API_KEY) {
		let err = 'ERROR: OPENAI_API_KEY unset\r\n';
		err += 'To set, use the command:\r\n';
		err += 'export OPENAI_API_KEY=<your_key>\r\n';
		err += 'https://platform.openai.com/account/api-keys';
		console.log(err);
		process.exit();
	}
}

function _loadOpts(options) {
	const content = fs.readFileSync(USER_CFG_DIR + '/settings.json', 'UTF-8');
	const optJSON = JSON.parse(content);
	options = Object.assign(optJSON, options);
	return options;
}

function _logHistory(ogPrompt, output) {
	const histNew = {
		user: ogPrompt,
		llm: output,
	};
	const historyJSON = _loadHistory(MAX_HISTORY_STORE, false, true);
	historyJSON.push(histNew);
	fs.writeFileSync(
		USER_CFG_DIR + '/history.json',
		JSON.stringify(historyJSON, null, 2)
	);
}

function _loadHistory(count = 1, reverse = true, json = false) {
	let content = fs.readFileSync(USER_CFG_DIR + '/history.json', 'UTF-8');
	if (!content) {
		console.log('WARNING: Can not read history file!');
		content = '[]';
	}
	let historyJSON = JSON.parse(content) || [];
	// Always reverse history first
	historyJSON.reverse();
	// Slice the history
	historyJSON = historyJSON.slice(0, count);
	// Undo reverse if needed
	if (reverse === false) {
		historyJSON.reverse();
	}
	if (json) {
		// Return json
		return historyJSON;
	}
	let historyStr = '';
	for (const item of historyJSON) {
		historyStr += '_user_: ' + item.user;
		historyStr += '\r\n';
		historyStr += '_llm_: ' + item.llm;
		historyStr += '\r\n';
	}
	return historyStr;
}

const program = new Command();

program
	.name('sllm')
	.description(
		'CLI for GPT3. Created by Mathieu Dombrock 2023. GPL3 License.'
	)
	.version(VERSION_NUMBER);

program
	.command('prompt', { isDefault: true })
	.description('send a prompt (default command)')
	.argument('<prompt...>', 'the prompt text')
	.option('-m, --max-tokens <number>', 'maximum tokens to use', '256')
	.option('-t, --temperature <number>', 'temperature to use', '0.2')
	.option('-c, --context <string...>', 'context to prepend')
	.option('-d, --domain <string...>', 'subject domain to prepend')
	.option('-e, --expert <string...>', 'act as an expert on this domain')
	.option('-5, --like-im-five', 'explain it like I\'m 5 years old')
	.option('-H, --history <number>', 'prepend history (chatGPT mode)')
	.option('-f, --file <path>', 'preprend the given file contents')
	.option('-v, --verbose', 'verbose output')
	.option('-M, --mock', 'dont actually send the prompt to the API')
	.action((prompt, options) => {
		llm(prompt, options);
	});

program
	.command('set')
	.description('set a persistant command option')
	.option('-m, --max-tokens <number>', 'maximum tokens to use', '256')
	.option('-t, --temperature <number>', 'temperature to use', '0.2')
	.option('-c, --context <string...>', 'dontext to prepend')
	.option('-d, --domain <string...>', 'subject domain to prepend')
	.option('-e, --expert <string...>', 'act as an expert on this domain')
	.option('-5, --like-im-five', 'explain it like I\'m 5 years old')
	.option('-H, --history <number>', 'prepend history (chatGPT mode)')
	.option('-f, --file <path>', 'preprend the given file contents')
	.option('-v, --verbose', 'verbose output')
	.option('-M, --mock', 'dont actually send the prompt to the API')
	.action((options) => {
		setOpts(options);
	});

program
	.command('settings')
	.description(
		'view the current settings that were changed via the `set` command'
	)
	.option('-d, --delete', 'Delete the current settings')
	.action((options) => {
		settings(options);
	});

program
	.command('hist')
	.description('manage the prompt / response history')
	.option('-v, --view <number>', 'view the prompt history', '8')
	.option('-d, --delete', 'delete the prompt history')
	.action((options) => {
		history(options);
	});

program
	.command('purge')
	.description('delete all history and settings')
	.action(() => {
		purge();
	});

program
	.command('count')
	.description('estimate the tokens used by a prompt or file')
	.option('-p, --prompt <string...>', 'the prompt string to check')
	.option('-f, --file <path>', 'the file path to check')
	.action((options) => {
		countTokens(options);
	});

program
	.command('repeat')
	.description('repeat the last response')
	.action(() => {
		repeat();
	});

program.parse();
